### DISCUSSION TITLE
deckgl properties: parent or container?

### DISCUSSION THREAD

**Discussion by fundef1** [visgl/deck.gl#5536] [Q&A]
Hi, I've figured out how to get it to work, but fail to understand what's happening.

simple testcase: page with a simpe ``` <div id="mapholder" ></div>```
i'm using ```   "https://unpkg.com/deck.gl@latest/dist.min.js" ``` as per the getting started docs
 
when creating deckgl with 
```
new DeckGL({
  ...blah...
  parent: 'mapholder'
})
```
It fails to properly render the map in the div.
however this works:
```
new DeckGL({
  ...blah...
  container: 'mapholder'
})
```

in the code i found: 
```
    if (!canvas) {
      canvas = document.createElement('canvas');
      const parent = props.parent || document.body;
      parent.appendChild(canvas);
    }
 ```
docs also mention parent as the correct property:
```
parent (HTMLElement)
   Default: document.body
The container to append the auto-created canvas to.
```

Can somebody explain what's going on, what i'm doing wrong?
thanks!

**Comment by sclebal** [comment]
Hi, `container` can be either a string or a DOMElement, but `parent` must be a DOMElement, you have to do document.getElementById.

References:

https://deck.gl/docs/api-reference/core/deckgl#container
https://deck.gl/docs/api-reference/core/deck#parent

**Comment by Pessimistress** [comment]
[container](https://deck.gl/docs/api-reference/core/deckgl#container) is only supported by the scripting interface (`deck.DeckGL` when you use the script tag) as a convenience. Under the hood it creates a mapbox instance (if Mapbox GL JS is imported via a script tag), a deck canvas and its tooltip inside the container.

[parent](https://deck.gl/docs/api-reference/core/deck#parent) is supported by all flavors of `Deck`. There is no built-in base map integration. If no `canvas` prop is specified, a canvas is created and attached to the parent. If you provide your own `canvas`, you can still specify a custom `parent` as the outer container that receives pointer events.

**Reply by fundef1** [reply]
thanks to you both for the quick respones. 
That's indeed how I thought it's supposed to work: from the docs:
   "All Deck class properties, with these additional props that can be passed to the constructor:"
yet I can't get the 'parent' to work as expected.

As an example I've added a div to the hexagon layer example on [codepen ](https://codepen.io/fundef/pen/LYbQxvO) 

mapboxgl-map is attached to body instead of mapholder.

**Reply by Pessimistress** [reply]
The scripting interface `DeckGL` has `container` default to `document.body` if you don't specify it. So that takes precedent of `parent`. The documentation should clarify this.
